docs/WritingTests.asciidoc

```diff
@@ -752,17 +768,169 @@ Sending key presses over VNC is very slow so for tests which send a lot of
 text commands it is much faster to use a serial port for both sending and
 receiving TTY commands.
 
+Communicating entirely using text also means that you no longer have to worry
+about your needles being invalidated due to a font change or similar. It is
+also much cheaper to transfer text and test it against regular expressions
+than encode images from a VNC feed and test them against sample images
+(needles).
+
+On the other hand you can no longer use +assert_screen+ or take a screen shot
+because the text is never rendered as an image. A lot of programs will also
+send ANSI escape sequences which will appear as raw text to the test script
+instead of being interpreted by a terminal emulator which then renders the
+text.
+
+However the advantages of using the serial terminal usually outweigh the
+negatives, if it is possible to use it at all.
+
 [source,perl]
 --------------------------------------------------------------------------------
 select_console('root-virtio-terminal');  # Selects a virtio based serial terminal
 --------------------------------------------------------------------------------
 
 Changing input and output to a serial terminal has the side effect of changing
-where +wait_serial+ reads output from. This will cause some distribution
-specific utility functions to fail, however they can usually be fixed with the
-+is_serial_terminal+ API function. To find out more look at the
-+is_serial_terminal+ POD in +testapi.pm+.
-
-Another consequence of moving to a serial terminal is that none of the needle
-based commands will be available because there is no screen image to match
-against.
+where +wait_serial+ reads output from. On a QEMU VM +wait_serial+ usually
+reads from an emulated serial port which is also where the kernel log is
+usually output to.
+
+When switching to a virtio based serial terminal, +wait_serial+ will then read
+from the virtio serial port instead. However the emulated serial port still
+exists and can receive output. Some utility library functions are hard coded
+to redirect output to the emulated serial port and expect that +wait_serial+
+will be able to read it. Usually it is not too difficult to fix the utility
+function, you just need to remove some redirection in the shell command.
+
+Another common problem is that some library or utility function tries to take
+a screen shot. The hard part is finding what takes the screen shot, but then
+it is just a simple case of checking +is_serial_terminal+ and not taking it if
+we are on a serial terminal.
+
+As mentioned previously, distributions tend to wrap +select_console+, so
+instead of using it directly, you can use something like
+
+[source,perl]
+--------------------------------------------------------------------------------
+if (select_virtio_console()) {
```

Comment:
it's a big question for me to document function which is in testing repository inside a tool . From openQA perspective it is just another testing repository. How guys from Fedora would follow this ? 
I am agree that select_virtio_console is needed function but to insert it to documentation need to move to os-autoinst

Reply:
Or at least you need to point where this function is declared and explain that it's only related to os-autoinst-distri-opensuse

Reply:
I prefer the latter because most distributions should have something similar and it is nice to have complete documentation in one place.